### DISCUSSION TITLE
TPSheets, SA Features and bugfixes

### DISCUSSION THREAD

**Discussion by Tirlititi** [Albeoris/Memoria#170] [General]
- Improved the TPSheet feature for atlas images (these images that contain many icons or UI elements). See https://github.com/Albeoris/Memoria/commit/73d39b56234fb94b84cd2959f11adca7effa5fd3#comments
- Improved Supporting Abilities feature system, mostly to allow features triggered by something else than a supporting ability. See https://github.com/Albeoris/Memoria/commit/3e2ef7eba91aaa3c9cfa6e5c72feb0f5e717cbf3#comments
- Bug fixes

<hr /><em>This discussion was created from the release <a href='https://github.com/Albeoris/Memoria/releases/tag/'>TPSheets, SA Features and bugfixes</a>.</em>

**Comment by Albeoris** [comment]
@Tirlititi Hi, is the format still compatible with TexturePacker?

**Reply by Tirlititi** [reply]
Hi @Albeoris 
I'm not sure of what kind of compatibility you want.

(1) Being able to import a .tpsheet generated by (or compatible with) the last format I did in Memoria into the TexturePacker tool.
(2) Being able for Memoria to use a .tpsheet generated by the TexturePacker tool.
(3) Being able for Memoria to use the specific format of .tpsheet that Memoria could read before I changed things.

About (1), it looks like TexturePacker can't import .tpsheet files, only export them, so unless I'm wrong, that's out of the question. We would need to generate .tps files instead (a kind of XML).
About (3), yes they are still compatible: .tpsheet files that could be used by Memoria beforehand can still be used without any change in behaviour.

About (2), that's the trickiest part because I can't seem to find any documentation for the format of .tpsheet...
I don't think that .tpsheet files that are generated by Memoria (when using "[Export] Graphics") are completly valid .tpsheet files. Most likely, it would require a better header to specify which number corresponds to what.

The .tpsheet that Memoria used to handle had to have the following format only:
```
:format=40000
:size=PNGSIZE
sprite_name;X;Y;WIDTH;HEIGHT;PIVOTX;PIVOTY
```

Now, .tpsheet can be of the format above but also of the following format:
```
:format=40000
:size=PNGSIZE
sprite_name;X;Y;WIDTH;HEIGHT;PIVOT_X;PIVOT_Y;PADDING_LEFT;PADDING_RIGHT;PADDING_TOP;PADDING_BOT;BORDER_LEFT;BORDER_RIGHT;BORDER_TOP;BORDER_BOT
```

I won't purchase the TexturePacker tool for testing how it works, but surely a correct format would look more like this:
```
:format=40000
:size=PNGSIZE
:pivotpoints=enabled
:paddings=enabled
:borders=enabled
sprite_name;X;Y;WIDTH;HEIGHT;PIVOT_X;PIVOT_Y;PADDING_LEFT;PADDING_RIGHT;PADDING_TOP;PADDING_BOT;BORDER_LEFT;BORDER_RIGHT;BORDER_TOP;BORDER_BOT
```
(Not sure about the keywords at all, especially "paddings".)

Also, note how FF9's sprites work: they don't use any pivot point in the class [UISpriteData](https://github.com/Albeoris/Memoria/blob/main/Assembly-CSharp/Global/UI/UISprite/UISpriteData.cs), so the PIVOT_X/PIVOT_Y values are discarded (they were already before I added padding and border).
